lib/bricks-evaluate.js

```diff
@@ -34,29 +58,170 @@ function createContext() {
+		var callWhen = b.callExpression(whenIdentifier, [
+			wrapWithWhenThen(memberExpression.object)
+		]);
+		
+		var thenMember = b.memberExpression(
+			callWhen,
+			thenIdentifier,
+			false
+		);
+		
+		var retIdentifier = b.identifier("ret");
+		var functionExpression = b.functionExpression(
+			null,
+			[ retIdentifier ],
+			b.blockStatement(
+				[
+					b.returnStatement(
+						b.memberExpression(
+							retIdentifier,
+							memberExpression.property,
+							memberExpression.computed
+						)
+					)
+				]
+			)
+		);
+		
+		var callThen = b.callExpression(thenMember, [
+			functionExpression
+		]);
+		
+		return callThen;
+	}
+	
+	function throwSyntaxError(node) {
+		throw new SyntaxError('Invalid expression: ' + recast.print(node).code);
+	}
+	
+	var visitors = {
+		visitMemberExpression: function (node) {
+			// Blacklist computed MemberExpression (e.g. obj["property"]).
+			if (node.value.computed) {
+				throwSyntaxError(node.value);
+			}
+			
+			node.replace(wrapWithWhenThen(node.value));
+			return false;
+		},
+		
+		// Blacklist JavaScript-specific expressions.
+		// @see https://github.com/benjamn/ast-types/blob/master/def/core.js
+		// @see Mozilla Parser API https://developer.mozilla.org/en-US/docs/Mozilla/Projects/SpiderMonkey/Parser_API
+		visitArrayExpression: throwSyntaxError,
+		visitObjectExpression: throwSyntaxError,
+		visitThisExpression: throwSyntaxError,
+		visitNewExpression: throwSyntaxError,
+		visitFunctionExpression: throwSyntaxError,
+		visitFunctionDeclaration: throwSyntaxError,
+		visitVariableDeclaration: throwSyntaxError,
+		visitVariableDeclarator: throwSyntaxError,
+		visitForInStatement: throwSyntaxError,
+		visitDebuggerStatement: throwSyntaxError
+	};
+	
+	var ast = recast.parse(code);
+	recast.visit(ast, visitors);
+	return recast.print(ast).code;
+}
+
 function evaluate(code, context, filename, callback) {
+	// Put a reference to `when` into the context.
+	// We generate random identifier to prevent exploiting it from the provided code.
+	var whenIdentifierName = '__when' + Math.floor(1000 + Math.random() * 1000);
+	context[whenIdentifierName] = when;
+	
+	var timing;
 	try {
+		// Convert the code to promise-based.
+		// `when` is referenced via the passed identifier.
+		code = transformCode(code, {
+			when: whenIdentifierName
+		});
+		
+		// Compile the script from the code.
 		var script = vm.createScript(code, {
 			filename: filename,
 			displayErrors: false
 		});
 		
+		// Start the timing measurement.
+		timing = require('./bricks-timing')();
+		
+		// Execute the script.
 		var result = script.runInContext(context, {
-			displayErrors: false
+			displayErrors: false,
+			timeout: 10000
 		});
 		
-		// Async evaluation.
+		// Wait for the result to resolve.
 		when(result).done(function (actualResult) {
-			callback(null, actualResult);
+			timing.end();
+			
+			// Treat `undefined` result as an error.
+			// TODO(sompylasar): Think up something for the APIs that return void.
```

Comment:
We won't need `Join()` in REPL.

As for `Register()/UnRegister()`, I was thinking of converting them to a scoped version (return a handle from `Register()`, automatically un-register once that handle goes out of scope). What do you think, in general and REPL-wise?